Add parsing to some single value aggregations

cbuescher opened the pull request:
Similar to #23973 this adds parsing from xContent for the Min, Max, Avg, Sum and ValueCount aggregation.
I tried to pull out common functionality to ParsedNumericMetricsAggregation.SingleValue. Since InternalCardinality 
and InternalValueCount store a long value instead of a double, parsing and getters/setters for the value are slightly 
different so I didn't include their parsed counterparts under ParsedNumericMetricsAggregation.SingleValue for now.
Also included one case where I tried pulling out common xContent rendering into the shared interface (in Sum) for discussion.
I'm unsure if I like that one but it would help reduce some code duplication.

This PR is WIP against a feature branch.

tlrx commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/max/ParsedMax.java:
```diff
@@ -0,0 +1,62 @@
+        return MaxAggregationBuilder.NAME;
+    }
+
+    private static final ObjectParser<ParsedMax, Void> PARSER = new ObjectParser<>(MaxAggregationBuilder.NAME, true, ParsedMax::new);
```
Can we move all statics down? (after doXContentBody())

cbuescher replied:
done

tlrx reviewed (commented):
I had a first look, I'm happy to see that many things can be mutualized. I don't really like how the parsed Cardinality/ValueCount extends the ParsedNumericMetricsAggregation.SingleValue and don't use the `double` there and provide their own `long` value. I'm wondering if we could let SingleValue take care of the stringified value and have a Long and Double version of SingleValue where all parsing logic and XContent rendering would be placed.

I have to admit that I'm not sure if that's possible at all, so I'll take a look at this too

javanna commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java:
```diff
@@ -0,0 +1,77 @@
+
+import java.io.IOException;
+
+public abstract class ParsedNumericMetricsAggregation extends ParsedAggregation {
```
do we need this class? and why does SingleValue have to be an inner class?

cbuescher replied:
I don't know, the idea was to pull out value()/getValueAsString() and the internal fields for the Min/Max/Sum/Avg aggregation, maybe some others of the single value aggregations that just store a double. I tried to mimic the inheritance structure of InternalNumericMetricsAggregation here, which contains inner classes for SingleValue and MultiValue. Maybe we don't need it, but I'd like to keep it for now and revisit this decision once we are done with all the single value aggregations.

javanna replied:
I think this hierarchy makes little sense and we should not copy it to our objects unless there are good reasons to do so. Is the only reason consistency?

cbuescher replied:
Is sharing value/valueAsString() a good reason to keep ParsedNumericMetricsAggregation? Or rename it to ParsedSingleValueMetricsAggregation and remove the inner SingleValue class? My suspicion at this point is that we might also reuse some stuff for the MultiValue case but I cannot tell until we start working on those.

javanna replied:
I do not see what ParsedNumericMetricsAggregation holds. Seems like it is used only as an ancestor for its inner class? In that case let's have ParsedSingleValueNumericMetricsAggregation as a top level class?

javanna commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java:
```diff
@@ -0,0 +1,77 @@
+            if (valueAsString != null) {
+                return valueAsString;
+            } else {
+                return DocValueFormat.RAW.format(value);
```
should we try and do without depending on DocValueFormat here? We can just do what this method does?

cbuescher replied:
I replaced this with Double.toString(value).

javanna commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java:
```diff
@@ -0,0 +1,62 @@
+        return AvgAggregationBuilder.NAME;
+    }
+
+    private static final ObjectParser<ParsedAvg, Void> PARSER = new ObjectParser<>(AvgAggregationBuilder.NAME, true, ParsedAvg::new);
```
I don't have a strong opinion here, I don't think it makes a whole lot of difference

cbuescher replied:
Will according to suggestion by @tlrx than, since I also don't mind much.

javanna commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/avg/ParsedAvg.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Override
+    protected XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        boolean hasValue = value != Double.POSITIVE_INFINITY;
```
where is this logic in InternalAvg?

cbuescher replied:
In InternalAvg the `value` field is only rendered if the avg normalizer (count) is not 0. (`count != 0 ? getValue() : null`). I choose to parse back `null` as Double.POSITIVE_INFINITY because that is what you get back when you divide a positive double by `0L`. So I check for that value here to get the same xContent output like InternalAvg.

javanna replied:
sounds good! leave a comment?

cbuescher replied:
I added a comment.

javanna commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/sum/Sum.java:
```diff
@@ -29,4 +33,13 @@
      */
     double getValue();
+
+    // TODO we could pull common xContent rendering into a static method in the interface
```
what's left todo here? the static method is already in the interface?

cbuescher replied:
This was an example of how we could share the xContent rendering logic between InternalXY and ParsedXY aggregations. I don't know if it would work in all places. I only did it for SUM, I could try to do it for Min/Max/Avg etc but that means slight changes in the existing internal aggregations. Not sure which way to go here atm, maybe you have some opinion?

javanna replied:
I think we should do it only when it's very simple to do. It is a bit funky to have this static method in the interface but that is the best place for it. @tlrx what do you think?

cbuescher replied:
I tried doing this for the aggregations in this PR in	1048678, but I'm not sure I like it. It avoids some code duplication but makes reading and reasonign about the code much harder in my opinion. Maybe we should only do it when it really saves a lot of duplication (like really long xContent methods), wdyt?

javanna replied:
yea I tend to agree, I am +0 on this :) I defer to Tanguy

javanna commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/valuecount/ParsedValueCount.java:
```diff
@@ -0,0 +1,75 @@
+    public String getValueAsString() {
+        // InternalValueCount doesn't print "value_as_string", but you can get a formated value using
+        // getValueAsString(). That method uses the raw formatter so we also use it here.
+        return DocValueFormat.RAW.format((double) valueCount);
```
here too, shall we try and not depend on the formatter?

cbuescher replied:
I changed this to Long.toString(valueCount).

javanna commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/valuecount/ParsedValueCount.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Override
+    public String getValueAsString() {
+        // InternalValueCount doesn't print "value_as_string", but you can get a formated value using
```
s/formated/formatted

cbuescher replied:
yes

javanna commented on core/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalMaxTests.java:
```diff
@@ -19,21 +19,31 @@
 import java.util.List;
 import java.util.Map;
 
+@Repeat(iterations=1)
```
leftover

cbuescher replied:
yes

javanna commented on core/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalMaxTests.java:
```diff
@@ -45,4 +55,24 @@ protected InternalMax createTestInstance(String name, List<PipelineAggregator> p
+            assertEquals(max.getValue(), parsed.getValue(), Double.MIN_VALUE);
+            assertEquals(max.getValueAsString(), parsed.getValueAsString());
+        } else {
+            // we write Double.NEGATIVE_INFINITY and Double.POSITIVE to xContent as 'null', so we
```
complete POSITIVE with POSITIVE_INFINITY ?

cbuescher replied:
done

javanna reviewed (approved):
left a few comments but LGTM

cbuescher commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java:
```diff
@@ -0,0 +1,77 @@
+            }
+        }
+
+        protected static void declareCommonSingeValueFields(
```
okay

cbuescher commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java:
```diff
@@ -0,0 +1,77 @@
+
+        protected static void declareCommonSingeValueFields(
+                ObjectParser<? extends ParsedNumericMetricsAggregation.SingleValue, Void> objectParser, double defaultNullValue) {
+            declareCommonFields(objectParser);
```
okay

cbuescher commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/ParsedNumericMetricsAggregation.java:
```diff
@@ -0,0 +1,77 @@
+                ObjectParser<? extends ParsedNumericMetricsAggregation.SingleValue, Void> objectParser, double defaultNullValue) {
+            declareCommonFields(objectParser);
+            objectParser.declareField(ParsedNumericMetricsAggregation.SingleValue::setValue,
+                    (parser, context) -> ParsedNumericMetricsAggregation.SingleValue.parseValue(parser, defaultNullValue),
```
sure

cbuescher commented:
@tlrx @javanna thanks for the reviews, I added changes according to most of your comments, left two discussions open for now before I squash/merge this later

tlrx commented:
> Maybe we should only do it when it really saves a lot of duplication (like really long xContent methods), wdyt?

Yes, I agree. I think we should only do it when it saves a lot of duplication with custom logic. In this case it saves 1 line but added many more, and it also "force" people to look at what the Sum.renderXContent() is doing. I think we should not add those.

cbuescher committed:
Adding parsing for InternalMax and InternalMin

cbuescher committed:
Adding parsing for InternalSum

cbuescher committed:
Adding parsing for InternalAvg

cbuescher committed:
Adding parsing for InternalValueCount

cbuescher committed:
Minor changes in ParsedCardinality

cbuescher commented:
@tlrx thanks, I reverted the commit where I pulled the xContent rendering into the interfaces, rebased and reordered the PR so changes are more self contained. I think I will merge this one manually to the feature branch without squashing so that the history better reflects the individual changes to each aggregation.
